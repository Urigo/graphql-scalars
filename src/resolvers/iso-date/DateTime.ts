/**
 * Copyright (c) 2017, Dirk-Jan Rutten
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { GraphQLScalarType, Kind } from 'graphql'
import type {GraphQLScalarTypeConfig} from 'graphql' // eslint-disable-line
import { validateJSDate, validateDateTime, validateUnixTimestamp } from './validator'
import { serializeDateTime, serializeDateTimeString, serializeUnixTimestamp, parseDateTime } from './formatter'

/**
 * An RFC 3339 compliant date-time scalar.
 *
 * Input:
 *    This scalar takes an RFC 3339 date-time string as input and
 *    parses it to a javascript Date.
 *
 * Output:
 *    This scalar serializes javascript Dates,
 *    RFC 3339 date-time strings and unix timestamps
 *    to RFC 3339 UTC date-time strings.
 */
const config: GraphQLScalarTypeConfig<Date, string> = {
  name: 'DateTime',
  description: 'A date-time string at UTC, such as 2007-12-03T10:15:30Z, ' +
               'compliant with the `date-time` format outlined in section 5.6 of ' +
               'the RFC 3339 profile of the ISO 8601 standard for representation ' +
               'of dates and times using the Gregorian calendar.',
  serialize (value) {
    if (value instanceof Date) {
      if (validateJSDate(value)) {
        return serializeDateTime(value)
      }
      throw new TypeError('DateTime cannot represent an invalid Date instance')
    } else if (typeof value === 'string') {
      if (validateDateTime(value)) {
        return serializeDateTimeString(value)
      }
      throw new TypeError(
        `DateTime cannot represent an invalid date-time-string ${value}.`
      )
    } else if (typeof value === 'number') {
      if (validateUnixTimestamp(value)) {
        return serializeUnixTimestamp(value)
      }
      throw new TypeError(
        'DateTime cannot represent an invalid Unix timestamp ' + value
      )
    } else {
      throw new TypeError(
        'DateTime cannot be serialized from a non string, ' +
        'non numeric or non Date type ' + JSON.stringify(value)
      )
    }
  },
  parseValue (value) {
    if (!(typeof value === 'string')) {
      throw new TypeError(
        `DateTime cannot represent non string type ${JSON.stringify(value)}`
      )
    }

    if (validateDateTime(value)) {
      return parseDateTime(value)
    }
    throw new TypeError(
      `DateTime cannot represent an invalid date-time-string ${value}.`
    )
  },
  parseLiteral (ast) {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `DateTime cannot represent non string type ${'value' in ast && ast.value}`
      )
    }
    const { value } = ast
    if (validateDateTime(value)) {
      return parseDateTime(value)
    }
    throw new TypeError(
      `DateTime cannot represent an invalid date-time-string ${String(value)}.`
    )
  }
}

export default new GraphQLScalarType(config)
