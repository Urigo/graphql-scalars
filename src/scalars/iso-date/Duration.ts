import { GraphQLScalarType, GraphQLScalarTypeConfig, Kind } from 'graphql';
import { createGraphQLError } from '../../error.js';

export type ISO8601Duration = string;

// original implementation
// https://stackoverflow.com/questions/32044846/regex-for-iso-8601-durations
// const ISO_DURATION_NEGATIVE_ALLOWED = /^-?P(?!$)(-?\d+(?:\.\d+)?Y)?(-?\d+(?:\.\d+)?M)?(-?\d+(?:\.\d+)?W)?(-?\d+(?:\.\d+)?D)?(T(?=-?\d)(-?\d+(?:\.\d+)?H)?(-?\d+(?:\.\d+)?M)?(-?\d+(?:\.\d+)?S)?)?$/
// const ISO_DURATION_WITHOUT_SIGN = /^P(?!$)(\d+(?:\.\d+)?Y)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?W)?(\d+(?:\.\d+)?D)?(T(?=\d)(\d+(?:\.\d+)?H)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?S)?)?$/

// negative and positive durations allowed, commas and decimal points valid for fractions
const ISO_DURATION =
  /^(-|\+)?P(?!$)((-|\+)?\d+(?:(\.|,)\d+)?Y)?((-|\+)?\d+(?:(\.|,)\d+)?M)?((-|\+)?\d+(?:(\.|,)\d+)?W)?((-|\+)?\d+(?:(\.|,)\d+)?D)?(T(?=(-|\+)?\d)((-|\+)?\d+(?:(\.|,)\d+)?H)?((-|\+)?\d+(?:(\.|,)\d+)?M)?((-|\+)?\d+(?:(\.|,)\d+)?S)?)?$/;

export const GraphQLDurationConfig: GraphQLScalarTypeConfig<string, string> = /*#__PURE__*/ {
  name: 'Duration',
  description: `
    A string representing a duration conforming to the ISO8601 standard,
    such as: P1W1DT13H23M34S
    P is the duration designator (for period) placed at the start of the duration representation.
    Y is the year designator that follows the value for the number of years.
    M is the month designator that follows the value for the number of months.
    W is the week designator that follows the value for the number of weeks.
    D is the day designator that follows the value for the number of days.
    T is the time designator that precedes the time components of the representation.
    H is the hour designator that follows the value for the number of hours.
    M is the minute designator that follows the value for the number of minutes.
    S is the second designator that follows the value for the number of seconds.

    Note the time designator, T, that precedes the time value.

    Matches moment.js, Luxon and DateFns implementations
    ,/. is valid for decimal places and +/- is a valid prefix
  `,

  serialize(value) {
    if (typeof value !== 'string') {
      throw createGraphQLError(`Value is not string: ${value}`);
    }

    if (!ISO_DURATION.test(value)) {
      throw createGraphQLError(`Value is not a valid ISO Duration: ${value}`);
    }

    return value;
  },

  parseValue(value) {
    if (typeof value !== 'string') {
      throw createGraphQLError(`Value is not string: ${value}`);
    }

    if (!ISO_DURATION.test(value)) {
      throw createGraphQLError(`Value is not a valid ISO Duration: ${value}`);
    }

    return value;
  },

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw createGraphQLError(`Can only validate strings as ISO Durations but got a: ${ast.kind}`, { nodes: ast });
    }
    if (!ISO_DURATION.test(ast.value)) {
      throw createGraphQLError(`Value is not a valid ISO Duration: ${ast.value}`, { nodes: ast });
    }

    return ast.value;
  },
  extensions: {
    codegenScalarType: 'string',
    jsonSchema: {
      title: 'Duration',
      type: 'string',
      pattern: ISO_DURATION.source,
    },
  },
};

export const GraphQLISO8601Duration: GraphQLScalarType = /*#__PURE__*/ new GraphQLScalarType({
  ...GraphQLDurationConfig,
  name: 'ISO8601Duration',
});

export const GraphQLDuration: GraphQLScalarType = /*#__PURE__*/ new GraphQLScalarType({
  ...GraphQLDurationConfig,
  name: 'Duration',
});
