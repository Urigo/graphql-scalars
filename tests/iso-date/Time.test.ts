/**
 * Copyright (c) 2017, Dirk-Jan Rutten
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { GraphQLTime } from '../../src/scalars/iso-date/Time';
import { Kind } from 'graphql';
// flowlint-next-line untyped-import:off
import MockDate from 'mockdate';
// flowlint-next-line untyped-import:off
import { stringify } from 'jest-matcher-utils';

// Mock the new Date() call so it always returns 2017-01-01T00:00:00.000Z
MockDate.set(new Date(Date.UTC(2017, 0, 1)));

const invalidDates = [
  'Invalid date',
  '2016-01-01T00:00:00.223Z',
  '10:30:02.Z',
  '00:00:00.45+0130',
  '00:00:00.45+01',
];

const validDates = [
  ['00:00:00Z', new Date(Date.UTC(2017, 0, 1))],
  ['00:00:59Z', new Date(Date.UTC(2017, 0, 1, 0, 0, 59))],
  ['10:30:02.1Z', new Date(Date.UTC(2017, 0, 1, 10, 30, 2, 100))],
  ['09:09:06.13Z', new Date(Date.UTC(2017, 0, 1, 9, 9, 6, 130))],
  ['10:00:11.003Z', new Date(Date.UTC(2017, 0, 1, 10, 0, 11, 3))],
  ['16:10:20.1359945Z', new Date(Date.UTC(2017, 0, 1, 16, 10, 20, 135))],
  ['00:00:00+01:30', new Date(Date.UTC(2016, 11, 31, 22, 30))],
  ['00:00:30.3-01:30', new Date(Date.UTC(2017, 0, 1, 1, 30, 30, 300))],
];

describe('GraphQLTime', () => {
  it('has a description', () => {
    expect(GraphQLTime.description).toMatchSnapshot();
  });

  describe('serialization', () => {
    [{}, [], null, undefined, true].forEach((invalidInput) => {
      it(`throws error when serializing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLTime.serialize(invalidInput),
        ).toThrowErrorMatchingSnapshot();
      });
    });

    // Serialize from Date
    [
      [new Date(Date.UTC(2016, 0, 1)), '00:00:00.000Z'],
      [new Date(Date.UTC(2016, 0, 1, 14, 48, 10, 3)), '14:48:10.003Z'],
    ].forEach(([value, expected]) => {
      it(`serializes javascript Date ${stringify(value)} into ${stringify(
        expected,
      )}`, () => {
        expect(GraphQLTime.serialize(value)).toEqual(expected);
      });
    });

    it(`throws error when serializing invalid date`, () => {
      expect(() =>
        GraphQLTime.serialize(new Date('invalid date')),
      ).toThrowErrorMatchingSnapshot();
    });

    [
      ['00:00:00Z', '00:00:00Z'],
      ['10:30:02.1Z', '10:30:02.1Z'],
      ['16:10:20.1359945Z', '16:10:20.1359945Z'],
      ['00:00:00+01:30', '22:30:00Z'],
      ['00:00:30.3-01:30', '01:30:30.3Z'],
    ].forEach(([input, output]) => {
      it(`serializes time-string ${input} into UTC time-string ${output}`, () => {
        expect(GraphQLTime.serialize(input)).toEqual(output);
      });
    });

    invalidDates.forEach((dateString) => {
      it(`throws an error when serializing an invalid date-string ${stringify(
        dateString,
      )}`, () => {
        expect(() =>
          GraphQLTime.serialize(dateString),
        ).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('value parsing', () => {
    validDates.forEach(([value, expected]) => {
      it(`parses date-string ${stringify(
        value,
      )} into javascript Date ${stringify(expected)}`, () => {
        expect(GraphQLTime.parseValue(value)).toEqual(expected);
      });
    });

    [4566, {}, [], true, null].forEach((invalidInput) => {
      it(`throws an error when parsing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLTime.parseValue(invalidInput),
        ).toThrowErrorMatchingSnapshot();
      });
    });

    invalidDates.forEach((dateString) => {
      it(`throws an error parsing an invalid time-string ${stringify(
        dateString,
      )}`, () => {
        expect(() =>
          GraphQLTime.parseValue(dateString),
        ).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('literial parsing', () => {
    validDates.forEach(([value, expected]) => {
      const literal = {
        kind: Kind.STRING,
        value: value.toString(),
      };

      it(`parses literal ${stringify(literal)} into javascript Date ${stringify(
        expected,
      )}`, () => {
        expect(GraphQLTime.parseLiteral(literal, {})).toEqual(expected);
      });
    });

    invalidDates.forEach((value) => {
      const invalidLiteral = {
        kind: Kind.STRING,
        value,
      };
      it(`errors when parsing invalid literal ${stringify(
        invalidLiteral,
      )}`, () => {
        expect(() =>
          GraphQLTime.parseLiteral(invalidLiteral, {}),
        ).toThrowErrorMatchingSnapshot();
      });
    });

    [
      {
        kind: Kind.FLOAT,
        value: '5',
      },
      {
        kind: Kind.DOCUMENT,
        // flowlint-next-line unclear-type:off
      } as any,
    ].forEach((literal) => {
      it(`errors when parsing invalid literal ${stringify(literal)}`, () => {
        expect(() => GraphQLTime.parseLiteral(literal, {})).toThrowError();
      });
    });
  });
});
