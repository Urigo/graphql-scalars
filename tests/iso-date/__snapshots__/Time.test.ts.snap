// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLTime has a description 1`] = `"A time string at UTC, such as 10:15:30Z, compliant with the \`full-time\` format outlined in section 5.6 of the RFC 3339profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar."`;

exports[`GraphQLTime literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "00:00:00.45+01"} 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+01."`;

exports[`GraphQLTime literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "00:00:00.45+0130"} 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+0130."`;

exports[`GraphQLTime literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "10:30:02.Z"} 1`] = `"Time cannot represent an invalid time-string 10:30:02.Z."`;

exports[`GraphQLTime literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2016-01-01T00:00:00.223Z"} 1`] = `"Time cannot represent an invalid time-string 2016-01-01T00:00:00.223Z."`;

exports[`GraphQLTime literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "Invalid date"} 1`] = `"Time cannot represent an invalid time-string Invalid date."`;

exports[`GraphQLTime serialization throws an error when serializing an invalid date-string "00:00:00.45+01" 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+01."`;

exports[`GraphQLTime serialization throws an error when serializing an invalid date-string "00:00:00.45+0130" 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+0130."`;

exports[`GraphQLTime serialization throws an error when serializing an invalid date-string "10:30:02.Z" 1`] = `"Time cannot represent an invalid time-string 10:30:02.Z."`;

exports[`GraphQLTime serialization throws an error when serializing an invalid date-string "2016-01-01T00:00:00.223Z" 1`] = `"Time cannot represent an invalid time-string 2016-01-01T00:00:00.223Z."`;

exports[`GraphQLTime serialization throws an error when serializing an invalid date-string "Invalid date" 1`] = `"Time cannot represent an invalid time-string Invalid date."`;

exports[`GraphQLTime serialization throws error when serializing [] 1`] = `"Time cannot be serialized from a non string, or non Date type []"`;

exports[`GraphQLTime serialization throws error when serializing {} 1`] = `"Time cannot be serialized from a non string, or non Date type {}"`;

exports[`GraphQLTime serialization throws error when serializing invalid date 1`] = `"Time cannot represent an invalid Date instance"`;

exports[`GraphQLTime serialization throws error when serializing null 1`] = `"Time cannot be serialized from a non string, or non Date type null"`;

exports[`GraphQLTime serialization throws error when serializing true 1`] = `"Time cannot be serialized from a non string, or non Date type true"`;

exports[`GraphQLTime serialization throws error when serializing undefined 1`] = `"Time cannot be serialized from a non string, or non Date type undefined"`;

exports[`GraphQLTime value parsing throws an error parsing an invalid time-string "00:00:00.45+01" 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+01."`;

exports[`GraphQLTime value parsing throws an error parsing an invalid time-string "00:00:00.45+0130" 1`] = `"Time cannot represent an invalid time-string 00:00:00.45+0130."`;

exports[`GraphQLTime value parsing throws an error parsing an invalid time-string "10:30:02.Z" 1`] = `"Time cannot represent an invalid time-string 10:30:02.Z."`;

exports[`GraphQLTime value parsing throws an error parsing an invalid time-string "2016-01-01T00:00:00.223Z" 1`] = `"Time cannot represent an invalid time-string 2016-01-01T00:00:00.223Z."`;

exports[`GraphQLTime value parsing throws an error parsing an invalid time-string "Invalid date" 1`] = `"Time cannot represent an invalid time-string Invalid date."`;

exports[`GraphQLTime value parsing throws an error when parsing [] 1`] = `"Time cannot represent non string type []"`;

exports[`GraphQLTime value parsing throws an error when parsing {} 1`] = `"Time cannot represent non string type {}"`;

exports[`GraphQLTime value parsing throws an error when parsing 4566 1`] = `"Time cannot represent non string type 4566"`;

exports[`GraphQLTime value parsing throws an error when parsing null 1`] = `"Time cannot represent non string type null"`;

exports[`GraphQLTime value parsing throws an error when parsing true 1`] = `"Time cannot represent non string type true"`;
